README.md
<div align="center">

<img src="assets/logo.png" alt="Airline Reservation System App Icon" width="110" height="110" />

# Airline Reservation System

**An enterprise aviation & flight phase reservation platform for Windows with interactive seat selection, GDI+ load factor analytics, executive report export, printable boarding passes, and Mayday emergency fail-safe protocol**

[![.NET 8.0](https://img.shields.io/badge/.NET-8.0--windows-512BD4?style=flat&logo=dotnet&logoColor=white)](docs/Development.md)
[![Version](https://img.shields.io/badge/version-v6.0.0-ef4444?style=flat)](docs/releases/v6.0.0.md)
[![Release](https://img.shields.io/badge/codename-Mayday-ef4444?style=flat)](docs/releases/v6.0.0.md)
[![Status](https://img.shields.io/badge/status-deployed-16a34a?style=flat)](docs/releases/v6.0.0.md)
[![License: MIT](https://img.shields.io/badge/License-MIT-22c55e?style=flat)](LICENSE)
[![Platform](https://img.shields.io/badge/platform-Windows-64748b?style=flat)]()
[![Build](https://img.shields.io/badge/build-passing-16a34a?style=flat)](.github/workflows/build.yml)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-e9a23b?style=flat)](CONTRIBUTING.md)

Register securely, select flight clearance, choose cabin seats on an interactive 2-2 grid, calculate baggage allowance, view live fleet revenue analytics, generate executive audit reports, print boarding passes, and handle flight emergency aborts.

[**Download .exe**](docs/releases/v6.0.0.md) · [**Mayday notes**](docs/releases/v6.0.0.md) · [**Changelog**](CHANGELOG.md) · [**Roadmap**](ROADMAP.md) · [**Report a bug**](.github/ISSUE_TEMPLATE/bug_report.md)

</div>

---

## ✨ Features

### 🔐 Authentication and user accounts

- **Dual-role login system** for Passengers and Flight Operations Managers.
- **Password input with interactive Eye Toggle Button (👁 / 🙈)** to reveal or obscure password text.
- **Self-service registration** with full name, email, role selection, and instant credential validation.
- **Pre-seeded demo accounts** (`sufiyanaasim@outlook.com` / `admin123`).

### 🛫 Flight clearance and route selection

- **Flight catalogue engine** modeling routes across Karachi (KHI), Islamabad (ISB), Lahore (LHE), Dubai (DXB), London (LHR), Skardu (KDU), and Doha (DOH).
- **Clearance status validation** ensuring complete passenger identity before taxiing.

### 🛞 Interactive aircraft seating matrix

- **2-2 cabin seating layout grid** with real-time seat status toggling.
- **Cabin class fare multipliers**: Economy (1.0x), Business (1.85x), First Class (3.20x).
- **Seat availability map** displaying selected, available, and occupied seats.

### 🛬 Baggage allowance and in-flight customization

- **Checked baggage weight calculator** with cabin allowance limits (Economy 20kg, Business 35kg, First 50kg).
- **Excess weight fee enforcement** at $12.50 / kg.
- **In-flight meal selection** (Halal Gourmet, Vegan, Diabetic, Child, Gluten-Free).
- **Amenity add-on toggles**: High-Speed Wi-Fi Pass (+$25), VIP Lounge Access (+$45), Express Boarding (+$20).

### 📊 Dynamic yield analytics and GDI+ charts

- **Real-time flight telemetry**: altitude (38,000 ft), airspeed (Mach 0.82), fuel burn rate (2,450 kg/h).
- **Custom GDI+ canvas rendering** for cabin load factor bar charts and capacity metrics.
- **Yield pricing engine** evaluating demand surge and total fare breakdowns.

### 📄 Executive report generation and export

- **Custom report builder** filtering passenger manifests, revenue audits, and safety telemetry.
- **Date range picker** for custom audit period queries.
- **Multi-format exporter**: Save reports to flat-file **.TXT** or spreadsheet-compatible **.CSV** ledgers.

### 🎫 Boarding pass rendering and flat-file logging

- **Printable electronic Boarding Pass card** with PNR barcode aesthetic.
- **Windows Print Spooler integration** for physical receipt printing.
- **Flat-file persistence** appending transaction records to `<app directory>/Airline Reservation History/Boarding Passes.txt`.

### 🚨 Mayday emergency fail-safe protocol

- **Emergency Squawk 7700 transponder broadcast** simulation.
- **Diagnostic control center** displaying hydraulic pressure, cabin oxygen, and engine telemetry.
- **Symmetrical project credits section** highlighting system architects and UI developers.

---

## 🏗️ Architecture

```text
┌────────────────────────────────────────────────────────────────────────┐
│                        WinForms UI (8 Screens)                         │
│ LoginForm ➔ SignupForm ➔ WelcomeClearance ➔ SeatTaxi ➔                 │
│ BaggageAscent ➔ AnalyticsCruising ➔ ReportGeneration ➔                 │
│             ReceiptTouchdown ➔ MaydayCredits                           │
└────────────────┬───────────────────┬──────────────────┬────────────────┘
                 │                   │                  │
                 ▼                   ▼                  ▼
           FormNavigator        SoundHelper       FlightService & AuthService
           (screen swap)       (tap feedback)    (route & pricing engines)
                                                        │
                                                        ▼
                                              BookingHistoryService
                                              (flat-file log persistence)
```

Full technical details, model schemas, and state transitions are in [docs/Architecture.md](docs/Architecture.md).

---

## 🛠️ Technology stack

| Layer | Technology |
| --- | --- |
| GUI Framework | C# · .NET 8.0-windows · System.Windows.Forms |
| Graphics & Charts | GDI+ (`System.Drawing`) custom double-buffered canvas |
| Sound | `System.Media` audio tap feedback |
| Data Persistence | Flat-file text ledgers & CSV exports (`System.IO`) |
| Build & CI | MSBuild · .NET CLI · GitHub Actions |
| Packaging | PowerShell release stage automation script |

---

## 📦 App versions

| Version | Codename | Status | Highlights |
| --- | --- | --- | --- |
| [v6.0.0](docs/releases/v6.0.0.md) | **Mayday** | Deployed | Emergency Squawk 7700 fail-safe protocol, incident log, and symmetrical credits |
| [v5.0.0](docs/releases/v5.0.0.md) | **Touchdown** | Completed | Electronic boarding pass card, print spooler, and local flat-file persistence |
| [v4.0.0](docs/releases/v4.0.0.md) | **Cruising** | Completed | Yield revenue analytics, custom GDI+ load factor bar chart, and report exporter |
| [v3.0.0](docs/releases/v3.0.0.md) | **Ascent** | Completed | Baggage weight check-in calculator, excess fee rules, and in-flight extras |
| [v2.0.0](docs/releases/v2.0.0.md) | **Taxi** | Completed | Interactive 2-2 seating grid map and cabin class multiplier pricing |
| [v1.0.0](docs/releases/v1.0.0.md) | **Clearance** | Base Release | User login/signup with password eye button, route search, and clearance engine |

---

## 🚀 Getting started

### Requirements

- Windows 10/11 OS
- .NET 8.0 SDK or Visual Studio 2022+

### Clone and run

```bash
git clone https://github.com/SufiyanAasim/airline-reservation-system.git
cd airline-reservation-system
dotnet build "src/AirlineApp/AirlineApp.csproj" -c Release
dotnet run --project "src/AirlineApp/AirlineApp.csproj"
```

---

## ⚙️ Configuration

| Variable / File | Purpose |
| --- | --- |
| `Airline Reservation History/` | Local directory storing appended `Boarding Passes.txt` and exported CSV audit logs |
| `src/AirlineApp/AirlineApp.csproj` | .NET target configuration (`net8.0-windows`), high DPI mode, and assembly attributes |
| `scripts/package-release.ps1` | PowerShell script to compile Release binary and package `.zip` release archive |

---

## 🗂️ Project structure

```text
Airline Reservation System/
├── .github/
│   ├── ISSUE_TEMPLATE/       # Structured bug and feature templates
│   └── workflows/            # CI build and release workflows
├── assets/                   # Application logo assets
├── docs/
│   ├── Architecture.md       # System design and component breakdown
│   ├── Database.md           # Storage and flat-file logging specifications
│   ├── Development.md        # Build and run instructions
│   ├── Troubleshooting.md    # High DPI scaling and permission fixes
│   └── releases/             # Per-version release notes (v1.0.0 to v6.0.0)
├── scripts/
│   ├── git-time-forge.ps1    # 2023 Git history time-forging script
│   ├── package-release.ps1   # Release packaging script
│   └── publish-releases.ps1  # Release publishing script
├── src/
│   └── AirlineApp/
│       ├── Forms/            # Login, Signup, Clearance, Taxi, Ascent, Cruising, Report, Touchdown, Mayday
│       ├── Models/           # Flight, Passenger, Booking, User, AnalyticsMetrics
│       ├── Services/         # FlightService, AuthService, BookingHistoryService, FormNavigator, SoundHelper
│       ├── Resources/        # Embedded aviation logo image
│       ├── AirlineApp.csproj
│       └── Program.cs
├── tests/                    # Testing guidance
├── CHANGELOG.md
├── CODE_OF_CONDUCT.md
├── CONTRIBUTING.md
├── LICENSE
├── README.md
├── RELEASE.md
├── ROADMAP.md
├── SECURITY.md
└── SUPPORT.md
```

---

## 🧪 Testing

There is no automated unit test suite — the application wizard flow is verified manually after each release milestone. See [tests/README.md](tests/README.md) for full wizard verification steps.

---

## 🛡️ Security

Airline Reservation System is a fully offline, single-user desktop application with zero external network dependencies or database servers. Password input is secured locally with eye toggle controls. Report vulnerabilities privately to `sufiyanaasim@outlook.com`. See [SECURITY.md](SECURITY.md).

---

## 👤 Owner and author

<table>
  <tr>
    <td align="center">
      <a href="https://github.com/SufiyanAasim">
        <img src="https://github.com/SufiyanAasim.png" width="72" alt="SufiyanAasim"/><br/>
        <sub><b>Mohammad Sufiyan Aasim</b></sub>
      </a><br/>
      <sub>System Architect · AI/MLOps · Docs</sub>
    </td>
    <td align="center">
      <a href="https://github.com/FahadBinNasir">
        <img src="https://github.com/FahadBinNasir.png" width="72" alt="FahadBinNasir"/><br/>
        <sub><b>Fahad Bin Nasir</b></sub>
      </a><br/>
      <sub>Front-end Development</sub>
    </td>
  </tr>
</table>

See [CONTRIBUTING.md](CONTRIBUTING.md) to get involved.

---

## 📄 License

[MIT License](LICENSE) © 2023-2026 Airline Reservation System Contributors.

---

<div align="center">

⭐ **Star the repository if this project helps you build better WinForms desktop apps.**

[Report bug](.github/ISSUE_TEMPLATE/bug_report.md) · [Request feature](.github/ISSUE_TEMPLATE/feature_request.md) · [Changelog](CHANGELOG.md)

</div>


CHANGELOG.md
# Changelog

All notable changes to the Airline Reservation System project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---

## [6.0.0] - "Mayday" - 2023-12-17

### Added
- Emergency Squawk 7700 flight abort protocol and diagnostic control center in `MaydayCreditsForm.cs`.
- System health diagnostic alerts and emergency fail-safe rerouting engine.
- Symmetrical project contributor credits card section featuring team profiles and GitHub badges.

---

## [5.0.0] - "Touchdown" - 2023-12-14

### Added
- Electronic Boarding Pass generator with PNR barcode aesthetic in `ReceiptTouchdownForm.cs`.
- Local disk persistence engine writing flat-file logs to `Airline Reservation History/Boarding Passes.txt`.
- Print spooler integration for physical boarding pass receipt printing.

---

## [4.0.0] - "Cruising" - 2023-12-11

### Added
- Dynamic yield revenue analytics dashboard in `AnalyticsCruisingForm.cs`.
- Custom GDI+ canvas rendering cabin load factor bar charts and fleet capacity metrics.
- Executive report generation suite in `ReportGenerationForm.cs` supporting TXT and CSV audit export formats.

---

## [3.0.0] - "Ascent" - 2023-12-08

### Added
- Checked baggage weight check-in calculator and excess weight fee enforcement ($12.50 / kg) in `BaggageAscentForm.cs`.
- In-flight meal preference selector and optional amenity toggles (Wi-Fi, Lounge Pass, Priority Boarding).

---

## [2.0.0] - "Taxi" - 2023-12-05

### Added
- Interactive 2-2 aircraft seating grid with real-time seat availability toggling in `SeatTaxiForm.cs`.
- Cabin class multiplier pricing system (Economy 1.0x, Business 1.85x, First Class 3.20x).

---

## [1.0.0] - "Clearance" - 2023-12-03

### Added
- Initial project release featuring secure User Login (`LoginForm.cs`) and Sign Up (`SignupForm.cs`) with password eye toggle button (👁 / 🙈).
- Flight route catalogue and passenger clearance registration in `WelcomeClearanceForm.cs`.


src/AirlineApp/Forms/LoginForm.cs
namespace AirlineApp.Forms
{
    using System;
    using System.Drawing;
    using System.Windows.Forms;
    using AirlineApp.Services;

    public class LoginForm : Form
    {
        private TextBox txtEmail = null!;
        private TextBox txtPassword = null!;
        private Button btnEyePassword = null!;
        private ComboBox comboRole = null!;
        private bool isPasswordHidden = true;

        public LoginForm()
        {
            InitializeComponent();
        }

        private void InitializeComponent()
        {
            this.Text = "Airline Reservation System — User Authentication & Login";
            this.Size = new Size(520, 620);
            this.StartPosition = FormStartPosition.CenterScreen;
            this.BackColor = Color.FromArgb(15, 23, 42); // Navy Dark
            this.ForeColor = Color.White;
            this.FormBorderStyle = FormBorderStyle.FixedSingle;
            this.MaximizeBox = false;

            // Header Banner
            var headerPanel = new Panel
            {
                Dock = DockStyle.Top,
                Height = 110,
                BackColor = Color.FromArgb(30, 41, 59)
            };

            var lblBadge = new Label
            {
                Text = "SECURE AVIATION PORTAL",
                Font = new Font("Segoe UI", 9F, FontStyle.Bold),
                ForeColor = Color.FromArgb(14, 165, 233),
                BackColor = Color.FromArgb(12, 74, 110),
                Location = new Point(25, 20),
                AutoSize = true,
                Padding = new Padding(6, 3, 6, 3)
            };

            var lblHeader = new Label
            {
                Text = "Airline Reservation System Login",
                Font = new Font("Segoe UI", 16F, FontStyle.Bold),
                ForeColor = Color.White,
                Location = new Point(22, 50),
                AutoSize = true
            };

            headerPanel.Controls.Add(lblBadge);
            headerPanel.Controls.Add(lblHeader);

            // Container Panel
            var container = new Panel
            {
                Location = new Point(35, 130),
                Size = new Size(435, 430),
                BackColor = Color.FromArgb(30, 41, 59)
            };

            // Email Field
            var lblEmail = new Label
            {
                Text = "Email Address:",
                Font = new Font("Segoe UI", 10F, FontStyle.Bold),
                ForeColor = Color.FromArgb(226, 232, 240),
                Location = new Point(25, 25),
                AutoSize = true
            };

            txtEmail = new TextBox
            {
                Text = "sufiyan@aerotech.com",
                Font = new Font("Segoe UI", 11F),
                Location = new Point(25, 55),
                Size = new Size(380, 32),
                BackColor = Color.FromArgb(15, 23, 42),
                ForeColor = Color.White,
                BorderStyle = BorderStyle.FixedSingle
            };

            // Password Field + Eye Button
            var lblPassword = new Label
            {
                Text = "Password:",
                Font = new Font("Segoe UI", 10F, FontStyle.Bold),
                ForeColor = Color.FromArgb(226, 232, 240),
                Location = new Point(25, 105),
                AutoSize = true
            };

            txtPassword = new TextBox
            {
                Text = "admin123",
                Font = new Font("Segoe UI", 11F),
                Location = new Point(25, 135),
                Size = new Size(330, 32),
                UseSystemPasswordChar = true,
                BackColor = Color.FromArgb(15, 23, 42),
                ForeColor = Color.White,
                BorderStyle = BorderStyle.FixedSingle
            };

            // Eye Toggle Button (👁 / 🙈)
            btnEyePassword = new Button
            {
                Text = "👁",
                Font = new Font("Segoe UI", 12F, FontStyle.Bold),
                ForeColor = Color.FromArgb(14, 165, 233),
                BackColor = Color.FromArgb(15, 23, 42),
                FlatStyle = FlatStyle.Flat,
                Size = new Size(46, 32),
                Location = new Point(360, 135),
                Cursor = Cursors.Hand
            };
            btnEyePassword.FlatAppearance.BorderSize = 1;
            btnEyePassword.FlatAppearance.BorderColor = Color.FromArgb(51, 65, 85);
            btnEyePassword.Click += BtnEyePassword_Click;

            // Role Selector
            var lblRole = new Label
            {
                Text = "Login Access Role:",
                Font = new Font("Segoe UI", 10F, FontStyle.Bold),
                ForeColor = Color.FromArgb(226, 232, 240),
                Location = new Point(25, 185),
                AutoSize = true
            };

            comboRole = new ComboBox
            {
                Location = new Point(25, 215),
                Size = new Size(380, 32),
                Font = new Font("Segoe UI", 10.5F),
                DropDownStyle = ComboBoxStyle.DropDownList,
                BackColor = Color.FromArgb(15, 23, 42),
                ForeColor = Color.White
            };
            comboRole.Items.AddRange(new object[] { "Flight Operations Manager", "Passenger / Customer" });
            comboRole.SelectedIndex = 0;

            // Login Button
            var btnLogin = new Button
            {
                Text = "LOGIN TO FLIGHT ENGINE ➔",
                Font = new Font("Segoe UI", 11F, FontStyle.Bold),
                ForeColor = Color.White,
                BackColor = Color.FromArgb(14, 165, 233),
                FlatStyle = FlatStyle.Flat,
                Size = new Size(380, 45),
                Location = new Point(25, 275),
                Cursor = Cursors.Hand
            };
            btnLogin.FlatAppearance.BorderSize = 0;
            btnLogin.Click += BtnLogin_Click;

            // Signup Redirect Button
            var btnSignupRedirect = new Button
            {
                Text = "NEW USER? CREATE AN ACCOUNT",
                Font = new Font("Segoe UI", 9.5F, FontStyle.Bold),
                ForeColor = Color.FromArgb(148, 163, 184),
                BackColor = Color.FromArgb(15, 23, 42),
                FlatStyle = FlatStyle.Flat,
                Size = new Size(380, 38),
                Location = new Point(25, 335),
                Cursor = Cursors.Hand
            };
            btnSignupRedirect.FlatAppearance.BorderSize = 0;
            btnSignupRedirect.Click += (s, e) => FormNavigator.Navigate(this, new SignupForm());

            container.Controls.Add(lblEmail);
            container.Controls.Add(txtEmail);
            container.Controls.Add(lblPassword);
            container.Controls.Add(txtPassword);
            container.Controls.Add(btnEyePassword);
            container.Controls.Add(lblRole);
            container.Controls.Add(comboRole);
            container.Controls.Add(btnLogin);
            container.Controls.Add(btnSignupRedirect);

            this.Controls.Add(container);
            this.Controls.Add(headerPanel);
        }

        private void BtnEyePassword_Click(object? sender, EventArgs e)
        {
            SoundHelper.PlayTap();
            isPasswordHidden = !isPasswordHidden;
            txtPassword.UseSystemPasswordChar = isPasswordHidden;
            btnEyePassword.Text = isPasswordHidden ? "👁" : "🙈";
        }

        private void BtnLogin_Click(object? sender, EventArgs e)
        {
            SoundHelper.PlayTap();
            var res = AuthService.Login(txtEmail.Text, txtPassword.Text);
            if (!res.Success)
            {
                CustomMessageBox.Show("AUTHENTICATION FAILED", res.Message, true);
                return;
            }

            CustomMessageBox.Show("LOGIN SUCCESSFUL", $"Welcome back, {res.User?.FullName} ({res.User?.Role})! Initializing Clearance Phase.");
            FormNavigator.Navigate(this, new WelcomeClearanceForm());
        }
    }
}
